### PR TITLE
Add `as_object_map` method to JsProxy

### PR DESCRIPTION
**Pull request by hoodmane**
This adds a method which returns a new JsProxy that acts like a container of the JavaScript object's `ownProperties`.

See #3267

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests
- [x] Add new / update outdated documentation

### CODE REVIEW

File: src/core/jsproxy.c [hunk 1]
```diff
@@ -1500,6 +1503,174 @@ static PyMethodDef JsProxy_finally_MethodDef = {
...
+  return result;
+}
+
+EM_JS_NUM(int, JsObjMap_length_js, (JsRef idobj), {
+  let jsobj = Hiwire.get_value(idobj);
+  let length = 0;
+  for (let _ of Module.iterObject(jsobj)) {
+    length++;
+  }
+  return length;
```
Comment (ryanking13) [inline src/core/jsproxy.c:1547]: Maybe we need to mention somewhere that it takes `O(n)` for `len(jsproxy_object_map)`?

**Review by ryanking13 — APPROVED** [review]
Thanks, @hoodmane! It looks really cool.

File: src/core/jsproxy.c (hunk 1 above)
Comment (hoodmane) [reply src/core/jsproxy.c:1547]: Yeah perhaps this is worth pointing out. This is a good reason to use a `Map` instead =)

**Comment by hoodmane** [comment]
Thanks for the review @ryanking13!

**Pull request merged**